### ISSUE TITLE
Unclear how to stop a client.invoke on a stream

### ISSUE THREAD

**Issue by delaneyj** [improbable-eng/grpc-web#66]
I have a working example of accessing a stream using `grpc.invoke` however there seems to be no way to cancel.  Would think there would be something like `setInterval` ala...

```
const cancelId = grpc.invoke(...)
grpc.cancelnvoke(cancelId)
```

thoughts?

**Comment by jonnyreeves** [comment]
@MarcusLongmuir how would you feel about a change in the API so that `invoke()` returns a client which allows the consumer to bind callback to messages (via a [Signal](https://github.com/millermedeiros/js-signals) based API) but also perform actions such as cancellation (which can be percolated through to the underlying transport).

The API would look something like this:

```js
const client = grpc.invoke(BookService.QueryBooks, {
  host: '...',
  request: '...'
});

// Bind handlers to the client via the Signal-based API.
handle.onHeaders.add(headers => { ... });
handle.onMessage.add(message => { ... });
handle.onEnd.add((code, message, trailers) => { ... });

// Now we can do fun things like cancel the request...
setTimeout(() => { 
  client.abort();
}, 100);
```

This could also be a nice way to clean up the API a little as we could drop the `unary()` method and instead provide a convince helper which wraps the `client` object, ie:

```js
const client = grpc.invoke( ... );
consumeResponse(client, result => { 
  const { status, statusMessage, headers, message, trailers } = result;
});
```

WDYT?

**Comment by MarcusLongmuir** [comment]
@jonnyreeves: One concern I had about this is that aborting `fetch` [isn't clearly defined](https://github.com/whatwg/fetch/issues/447), but it appears that this [works with ReadableStream](https://github.com/whatwg/fetch/issues/447#issuecomment-313236964) (which is [required for grpc-web-client](https://github.com/improbable-eng/grpc-web/blob/3dc38343649003c9e2e187efbe7e61ccb08bf960/ts/src/transports/Transport.ts#L65) to use the browser's `fetch` implementation).

I see the reason behind changing the API - it forces the return value of `grpc.invoke` to be available in all contexts to avoid an issue like:
```ts
const client = grpc.invoke(BookService.QueryBooks, {
  host: '...',
  request: '...',
  onHeaders: headers => {
    // client might not be defined if onHeaders is called immediately by .invoke()
    client.abort();
  }
});
```

If this is the only reason to change the API then maybe there's a less impactful way to get this aborting functionality? At a minimum we can just have `.invoke` return an object with a `.abort()` function, but it could introduce the above issue.

If you think that we should change the API regardless of the addition of aborting then it might be best to create another issue to discuss the pros and cons of the approaches we can take and link it back to this issue.

**Comment by jonnyreeves** [comment]
@MarcusLongmuir You're right - I thought about this further and binding the handlers to a client object in this fashion is going to cause a potential race condition where the RPC could complete before the handlers are bound - we could cache response data but this could very easily lead to memory leaks.

I still think it makes sense to return an object over a func just incase we want to add additional methods / data in future - I've whipped up some code that I *think* works, just need to write a test for it then I'll raise a PR.

**Comment by delaneyj** [comment]
@jonnyreeves  that API works for me!

**Comment by plaisted** [comment]
Couldn't you get around the issues by calling invoke() on the client instead of it automatically being requested?

```typescript
const client = grpc.create(BookService.QueryBooks, {
  host: '...',
  request: '...'
});

// Bind handlers to the client via the Signal-based API.
client.onHeaders.add(headers => { ... });
client.onMessage.add(message => { ... });
client.onEnd.add((code, message, trailers) => { ... });
// start 
client.Invoke();

// Now we can do fun things like cancel the request...
setTimeout(() => { 
  client.abort();
}, 100);
```

Also talking about the API it would be nice to have a version of the unary library return a promise instead of it's callback setup:
```typescript
const getBookRequest = new GetBookRequest();
getBookRequest.setIsbn(60929871);
grpc.unary(BookService.GetBook, {
  debug: false,// optional - enable to output events to console.log
  request: getBookRequest,
  host: host
}).then(res => {
    const { status, statusMessage, headers, message, trailers } = res;
    if (status === Code.OK && message) {
      console.log("all ok. got book: ", message.toObject());
    }
  }
);
```

That would allow all the benefits of promises (error catching, await, Promise.All, etc).

**Comment by MarcusLongmuir** [comment]
This has been released in `grpc-web-client@0.3.1`

**Comment by lozandier** [comment]
@plaisted I'd say it'd probably be better to look into leveraging an (ESNext) `Observable` to have better experiences managing async behavior than Promise abstractions can leveraging things like `Observable.merge(), Observable.subscribe()` & so on.
